### PR TITLE
[GEOS-9296] Fix race condition in audit logger initialization

### PR DESCRIPTION
**Pull request by sikeoka**
This pull request synchronizes the audit logger initialization to fix a race condition that could cause the audit log files to be written to the wrong directory. https://osgeo-org.atlassian.net/browse/GEOS-9296

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [X] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [X] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [X] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [X] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [X] PR for bug fixes and small new features are presented as a single commit
- [X] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [X] New unit tests have been added covering the changes
- [X] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [X] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [X] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by sikeoka** [comment]
@aaime, any chance you could take a look at this PR please?  Thank you.

File: src/extension/monitor/core/src/main/java/org/geoserver/monitor/auditlog/AuditLogger.java [hunk 1]
```diff
@@ -80,8 +80,8 @@ public AuditLogger(MonitorConfig config, GeoServerResourceLoader loader) throws
         templateConfig.setTemplateLoader(new AuditTemplateLoader(loader));
     }
 
-    void initDumper() throws IOException {
-        if (getProperty("enabled", Boolean.class, false)) {
+    synchronized void initDumper() throws IOException {
+        if (this.dumper == null && getProperty("enabled", Boolean.class, false)) {
```
Comment (aaime) [inline src/extension/monitor/core/src/main/java/org/geoserver/monitor/auditlog/AuditLogger.java]: Doh, this makes for a working double checked locking, but one that is less obvious to spot... maybe best to do the usual pattern in the caller method, "requestPostProcessed" instead?, .e.g:

```
   if (dumper == null) {
      synchronized(this) {
          if (dumper == null) initiDumper();
      }
```

File: src/extension/monitor/core/src/test/java/org/geoserver/monitor/auditlog/AuditLoggerTest.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    private File bad;
+    private File good;
+    private GeoServerResourceLoader loader;
+    private MonitorConfig config;
+    private GenericApplicationContext context;
+
+    @Before
+    public void setUp() throws Exception {
+        this.bad = this.directory.newFolder("bad");
```
Comment (aaime) [inline src/extension/monitor/core/src/test/java/org/geoserver/monitor/auditlog/AuditLoggerTest.java]: I don't understand this one... even if two threads end up performing the initialization, won't they both stop when they [find the system variable](https://github.com/geoserver/geoserver/blob/a38bc435cb9889f60a84fb67c815f5ad6c218035/src/extension/monitor/core/src/main/java/org/geoserver/monitor/auditlog/AuditLogger.java#L86) and end up ignoring the "bad" config value? The config field is used only if the path retrieved from the system property is null or empty, I don't see code resetting it to null or empty.

I would expect the race condition to have two dumpers created against the same directory and end up overwriting the first few request worth of logs, likely causing the log files to be invalid.

File: src/extension/monitor/core/src/main/java/org/geoserver/monitor/auditlog/AuditLogger.java (hunk 1 above)
Comment (sikeoka) [reply src/extension/monitor/core/src/main/java/org/geoserver/monitor/auditlog/AuditLogger.java]: There are two places in requestPostProcessed where initDumper is called.  It seems a little messy to have to do that synchronized code twice.

File: src/extension/monitor/core/src/test/java/org/geoserver/monitor/auditlog/AuditLoggerTest.java (hunk 2 above)
Comment (aaime) [reply src/extension/monitor/core/src/test/java/org/geoserver/monitor/auditlog/AuditLoggerTest.java]: No feedback on this comment?

Comment (sikeoka) [reply src/extension/monitor/core/src/test/java/org/geoserver/monitor/auditlog/AuditLoggerTest.java]: There are three possible bad effects of this race condition that I've seen.

1. What you mentioned where multiple RequestDumper threads will be created that try to write to the same log file.  Only one thread will succeed so a few requests will not be logged and there may be a few extra RequestDumper threads that will never exit.  This is not dependent on the system property and is self-correcting other than a minor leak.
2. Multiple RequestDumper threads will be created that try to write to different log files.  A few audit log files may be created with only a single request in them and there may be a few extra RequestDumper threads that will never exit.  This is not dependent on the system property and is self-correcting other than a minor leak.
3. The most severe bug is dependent on the system property and will not automatically correct itself.  The only viable option to fix this with a clustered GeoServer environment using a shared data directory is to restart any bugged GeoServer nodes until they are in a good state.  initDumper() sets the AuditLogger path to the system property, creates the resource from the path, and then sets the path to the property from the config file.  When multiple threads call initDumper() at the same time, it is possible for one thread to set the path to the config file value between the other thread setting it to the system property and creating the resource.

File: src/extension/monitor/core/src/main/java/org/geoserver/monitor/auditlog/AuditLogger.java (hunk 1 above)
Comment (sikeoka) [reply src/extension/monitor/core/src/main/java/org/geoserver/monitor/auditlog/AuditLogger.java]: Do you have any additional thoughts about this @aaime?

File: src/extension/monitor/core/src/test/java/org/geoserver/monitor/auditlog/AuditLoggerTest.java (hunk 2 above)
Comment (sikeoka) [reply src/extension/monitor/core/src/test/java/org/geoserver/monitor/auditlog/AuditLoggerTest.java]: My unit test checks for all of these possible error conditions.

**Comment by aaime** [comment]
I still don't understand the race condition, but I guess I won't get enough time and peace to look at it, I have no working hours  to look at it, and during the weekend the house is noisy. Going to trust you on this one and merging.

**Pull request merged**